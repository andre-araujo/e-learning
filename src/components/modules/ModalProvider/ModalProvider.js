import { Component, createContext } from 'react';
import Router from 'next/router';
import { node } from 'prop-types';
import { FaTimes } from 'react-icons/fa';

import {
  Container,
  Outer,
  Inner,
  Header,
  Content,
  Button,
  Close,
  Children,
} from './styles';

const ModalContext = createContext();

class ModalProvider extends Component {
  static Toggle = ({
    children, className, render, title, fullScreen,
  }) => (
    <ModalContext.Consumer>
      {data => (
        <Button
          type="button"
          aria-label="toggle modal"
          onClick={() => {
            data.toggleModal();
            data.setRenderComponent(render);
            data.setTitle(title);
            data.setFullScreen(fullScreen);
          }}
          className={className}
        >
          {children}
        </Button>
      )}
    </ModalContext.Consumer>
  )

  state = {
    isOpen: false,
    render: null,
    title: '',
    fullScreen: false,
  };

  componentDidMount() {
    Router.onRouteChangeComplete = () => {
      const { isOpen } = this.state;

      if (isOpen) {
        this.setState({
          isOpen: false,
          render: null,
          title: '',
          fullScreen: false,
        });
      }
    };
  }

  componentWillUnmount() {
    Router.onRouteChangeComplete = undefined;
  }

  setRenderComponent = (render) => {
    this.setState({ render });
  }

  setTitle = (title) => {
    this.setState({ title });
  }

  setFullScreen = (fullScreen) => {
    this.setState({ fullScreen });
  }

  toggleModal = () => {
    this.setState(({ isOpen }) => ({ isOpen: !isOpen }));
  };

  render() {
    const {
      toggleModal, setRenderComponent, setTitle, setFullScreen,
    } = this;
    const { children } = this.props;
    const {
      isOpen, render, title, fullScreen,
    } = this.state;

    return (
      <ModalContext.Provider value={
        {
          isOpen,
          toggleModal,
          setRenderComponent,
          setTitle,
          setFullScreen,
        }
      }
      >
        <Children isOpen={isOpen}>
          {children}
        </Children>
        {isOpen && (
          <Container aria-label="oppened modal" isOpen={isOpen} fullScreen={fullScreen}>
            <Outer>
              <Inner>
                <Header>
                  {title}
                  <ModalProvider.Toggle>
                    <Close>
                      <FaTimes />
                    </Close>
                  </ModalProvider.Toggle>
                </Header>
                <Content>
                  {typeof render === 'function' ? render({ isOpen, toggleModal }) : render}
                </Content>
              </Inner>
              <ModalProvider.Toggle className="overlay" />
            </Outer>
          </Container>
        )}
      </ModalContext.Provider>
    );
  }
}

ModalProvider.propTypes = {
  children: node.isRequired,
};

export default ModalProvider;
