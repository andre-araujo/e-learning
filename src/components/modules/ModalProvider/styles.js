import styled, { css } from 'styled-components';

export const Container = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  overflow: auto;
  overflow-x: hidden;
  overflow-y: auto;
  transform: translate(0,0);
  display: none;

  ${({ isOpen }) => isOpen && css`
    display: block;
  `}

  .overlay {
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    background: rgba(0, 0, 0, 0.4);
    cursor: pointer;
    padding: 0;
    display: block;
    width: 100vw;
    height: 100vh;
    border: 0;
  }
`;

export const Outer = styled.section`
  position: relative;
  transform: translate(0,0);
  display: flex;
  position: relative;
  align-items: center;
  justify-content: center;
  min-height: calc(100% - 60px);
  flex-direction: column;
  padding: 30px;
`;

export const Inner = styled.div`
  position: relative;
  z-index: 2;
`;

export const Header = styled.header`
  line-height: 1;
  height: 50px;
  background: #eee;
  z-index: 1;
  display: flex;
  justify-content: space-between;
  padding: 15px;
  padding-right: 40px;
  border-radius: 10px 10px 0 0;
  position: relative;
  font-size: 20px;
`;

export const Content = styled.div`
  z-index: 1;
  line-height: 1;
  width: 100%;
  background: #fafafa;
  padding: 15px;
  border-radius: 0 0 10px 10px;
`;

export const Button = styled.span`
  cursor: pointer;
  background: none;
  border: none;
`;

export const Close = styled.i`
  color: #333;
  position: absolute;
  right: 5px;
  top: 50%;
  padding: 5px;
  font-size: 20px;
  transform: translateY(-50%);

  :hover {
    color: orange;
  }
`;

export const Children = styled.div`
  transition: .2s filter;

  ${({ isOpen }) => isOpen && css`
    filter: blur(2px);
  `}
`;
