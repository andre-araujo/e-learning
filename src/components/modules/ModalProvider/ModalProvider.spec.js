import React from 'react';
import { mount } from 'enzyme';

import ModalProvider from './ModalProvider';

describe('ModalProvider component', () => {
  it('should match snapshot', () => {
    expect(mount(<ModalProvider />)).toMatchSnapshot();
  });
});
