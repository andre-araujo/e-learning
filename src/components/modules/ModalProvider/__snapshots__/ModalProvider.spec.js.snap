// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalProvider component should match snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "isOpen": false,
      "setFullScreen": [Function],
      "setRenderComponent": [Function],
      "setTitle": [Function],
      "toggleModal": [Function],
    }
  }
>
  <styled.div
    isOpen={false}
  >
    <div>
      children
    </div>
  </styled.div>
</ContextProvider>
`;
